### PR TITLE
feat: Feature/security-and-migration-enhancements

### PR DESCRIPTION
**Pull request by kpeacocke**
* fix(security): harden Chef server UI and update base image

* chore(deps): bump Python base image to 3.13.11 for OS-level CVE remediation

- Update Dockerfile Python version from 3.13.2 to 3.13.11
- Update release.yml build args to match latest stable Python 3.13.x
- Includes 9 patch versions worth of security updates in base image
- Addresses OS-level package vulnerabilities (libc, curl, zlib, util-linux, etc.)
- Maintains slim variant for minimal attack surface

Testing:
- All 1723 tests passing
- linting check: OK

* feat(migration): add interactive configuration and questionnaire for Chef to Ansible migration

* feat(assessment): add activity-level breakdown with time projections

- Add ActivityBreakdown dataclass for detailed effort tracking
- Calculate per-activity effort estimates (recipes, templates, attributes, etc.)
- Show manual vs AI-assisted hours/days for each component type
- Include complexity multipliers and efficiency gains per activity
- Display time savings and percentage reductions
- Integrate breakdown into cookbook assessment reports

Activity types covered:
- Recipes: 3h/item manual → 35% with AI (save 65%)
- Templates: 2h/item manual → 40% with AI (save 60%)
- Attributes: 1.5h/item manual → 30% with AI (save 70%)
- Custom Resources: 5h/item manual → 55% with AI (save 45%)
- Libraries: 4h/item manual → 60% with AI (save 40%)
- Handlers: 1h/item manual → 25% with AI (save 75%)
- Files: 0.5h/item manual → 20% with AI (save 80%)
- Definitions: 2.5h/item manual → 45% with AI (save 55%)

Testing:
- All 1723 tests passing
- Linting: OK

* test: Add comprehensive unit tests for activity breakdown and migration config

- Add 15 unit tests for ActivityBreakdown dataclass and calculation function
- Add 9 unit tests for MigrationConfig serialization and enumerations
- Test all ActivityBreakdown properties (manual_days, ai_assisted_days, time_saved_hours, efficiency_gain_percent)
- Test MigrationConfig serialization roundtrip (to_dict/from_dict)
- Test JSON serialization/deserialization
- Test enum value mappings (DeploymentTarget, MigrationStandard, ValidationTool)
- Use pytest.approx() for all float comparisons to avoid SonarQube warnings
- Fix metric key names (resource_count, attributes_count) to match implementation
- Fix string representation format expectations (1 decimal rounding)
- All 24 new tests passing, bringing total to 1747 tests

* feat(cli): add configure-migration command for interactive and CLI-based configuration

- Add new 'configure-migration' command to CLI with multiple options
- Support both interactive questionnaire mode (--interactive/-i flag)
- Support non-interactive mode with CLI arguments (deployment-target, migration-standard, etc.)
- Allow saving configuration to JSON file (--output option)
- Import MigrationConfig, DeploymentTarget, MigrationStandard, ValidationTool from migration_config module
- Add comprehensive CLI tests covering all use cases:
  * Non-interactive with single argument
  * Multiple CLI arguments
  * Output to file
  * Multiple validation tools
  * All options combined
- All 1752 tests passing (added 5 new CLI tests)
- Support for:
  * --deployment-target: app/awx/aap/native
  * --migration-standard: standard/flat/hybrid
  * --inventory-source: Chef Server or other
  * --validation-tools: Can specify multiple tools
  * --python-version: Target Python version
  * --ansible-version: Target Ansible version

Examples:
  souschef configure-migration --interactive  # Run questionnaire
  souschef configure-migration --deployment-target awx --validation-tools ansible-lint
  souschef configure-migration -i --output config.json  # Save to file

* feat(ui): add migration configuration and activity breakdown visualisation

- Created comprehensive migration config UI page with interactive forms
- Added activity breakdown visualisation with metrics and charts
- Implemented configuration options: deployment target, migration standard,
  inventory source, validation tools, Python/Ansible versions
- Added activity metrics display: manual hours, AI-assisted hours, time saved,
  efficiency gain percentage, timeline estimate
- Implemented export functionality: JSON and Markdown formats
- Added public API: calculate_activity_breakdown() in assessment.py
- Integrated migration config page into main navigation (8 buttons)
- Created UI pages package with __init__.py for proper imports
- Added 5 unit tests for configuration and activity breakdown functions

Features:
- Interactive configuration form with dropdowns and multiselect
- Activity breakdown table with calculated time savings
- Time savings visualisation with pandas bar chart
- Summary metrics display with 4 key metrics
- One-click export to JSON or Markdown formats
- Session state persistence for user configurations

Technical changes:
- souschef/ui/pages/migration_config.py: New UI page (~400 lines)
- souschef/ui/app.py: Navigation integration (8 buttons)
- souschef/assessment.py: Public API calculate_activity_breakdown()
- souschef/ui/pages/__init__.py: Package initialization
- tests/unit/test_migration_config_ui.py: 5 unit tests

All tests passing: 1757 passed, 1 skipped
Quality checks: ✓ Ruff ✓ Mypy ✓ Pytest

* feat(ui): enhance activity breakdown with detailed descriptions and improved layout

- Added count and description fields to activity breakdown output
- Implemented two-column layout: activity summary (left) and detail table (right)
- Activity summary displays card-like format with:
  - Activity name and item count
  - Description in italics
  - Manual vs AI-assisted hours comparison
  - Time saved and efficiency percentage
- Updated detail table to include Count and Description columns
- Fixed target Python version display in config summary
- Enhanced test to validate new activity fields
- Improved Markdown export with activity descriptions

Layout improvements:
- Summary metrics at top (4 columns)
- Activity summary in left column (scannable cards)
- Detailed table in right column (full data with 7 columns)
- Visual separation with dividers between activities

All tests passing: 5/5 migration config UI tests
Quality checks: ✓ Ruff ✓ Mypy

* fix(database): ensure analysis_results table persists before subsequent operations

- Added conn.commit() after CREATE TABLE IF NOT EXISTS analysis_results
- Added conn.commit() after CREATE TABLE IF NOT EXISTS conversion_results
- Removed redundant ALTER TABLE for cookbook_blob_key (already defined in CREATE TABLE)
- Prevents UndefinedTable error when History page initializes database

The issue was that table creation wasn't being committed before trying to
create indexes on it, causing psycopg.errors.UndefinedTable when accessing
the history page in the UI.

* feat(history): display activity breakdown details in analysis view

- Added _display_analysis_activity_breakdown() function to show activity details
- Displays two-column layout: activity summary (left) and details table (right)
- Activity summary shows item count, description, manual/AI hours, and time saved
- Details table includes all metrics in structured format
- Extracted from analysis_data JSON for each analysis result
- Provides same rich breakdown view as migration config page
- Fixes missing assessment details in history view

* fix(analysis): persist activity breakdown in analysis_data for history retrieval

- Added _serialize_activity_breakdown() helper function
- Serializes ActivityBreakdown objects to JSON-compatible dicts
- Includes activity_breakdown in analysis_data when saving to database
- Stores: activity_type, count, manual_hours, ai_assisted_hours,
  complexity_factor, description, time_saved_hours, efficiency_gain_percent
- Ensures activity breakdown is available when viewing analysis history
- Fixes missing activity details in saved analyses

* feat(analysis): display activity breakdown in cookbook analysis results

- Added _display_cookbook_activity_breakdown() function
- Displays two-column layout: activity summary (left) and details table (right)
- Handles both dict and ActivityBreakdown object formats
- Shows for each activity: name, count, description, manual/AI hours, time saved
- Details table with all metrics in structured format
- Activity breakdown now visible immediately after analysis completes
- Matches layout used in History and Migration Config pages

* fix(assessment): include activity_breakdown in assessment function returns

The root cause: Display functions were looking for activity_breakdown but
assessment functions weren't returning it.

Changes:
- assess_single_cookbook_with_ai(): Added activity_breakdown to return dict
- parse_chef_migration_assessment(): Added activity_breakdown from first cookbook
- Both functions now include estimated_hours_with_souschef for consistency
- Activity breakdown data now flows through to UI display functions

This fixes the missing activity details in both analysis results and history.

* fix(assessment): calculate activity_breakdown in AI assessment path

The AI assessment function _assess_single_cookbook_with_ai() was missing
the activity_breakdown calculation that exists in the rule-based version.
This caused activity breakdown data to not flow through to the UI even
though display functions were implemented.

Changes:
- Add _calculate_activity_breakdown() call in _assess_single_cookbook_with_ai()
- Include activity_breakdown in assessment dict returned by AI path
- Now matches the rule-based _assess_single_cookbook() behavior

This completes the data pipeline: assessment -> display -> database -> history

* feat: replace deprecated use_container_width and add history deletion

Changes:
1. Replace deprecated Streamlit use_container_width with width parameter
   - Updated all navigation buttons to use width='stretch'
   - Updated all dataframes to use width='stretch'
   - Updated other buttons throughout UI

2. Add delete functionality to history management
   - Added delete_analysis() and delete_conversion() methods to both StorageManager and PostgresStorageManager
   - Delete cascades: deleting analysis also deletes associated conversions
   - Returns boolean success/failure status

3. UI: Add delete buttons to History page
   - Analysis details now includes Delete Analysis button with warning
   - Conversion details now includes Delete Conversion button with warning
   - Both trigger immediate UI refresh on success

4. CLI: Add history management commands
   - souschef history list: List analysis and conversion history with filters
   - souschef history delete: Delete analysis or conversion by ID
   - Includes --yes flag to skip confirmation prompts

All quality checks passing:
- Ruff: All checks passed
- Mypy: Success, no issues found
- Tests: 1757 passed, 1 skipped

* feat: add history selection to Migration Planning, Dependency Mapping, and Validation Reports

Changes:
1. Migration Planning page:
   - Added 'Load from History' section to select from previous analyses
   - Pre-populates cookbook paths from historical analysis
   - Allows users to quickly reuse analyzed cookbooks

2. Dependency Mapping page:
   - Added 'Load from History' section
   - Added 'Use History' option to input method radio buttons
   - Loads cookbook path from session state when selected

3. Validation Reports page:
   - Added dual-tab history selection (Analysis and Conversion)
   - Analysis tab: Load cookbook path from previous analyses
   - Conversion tab: Load playbook path from previous conversions
   - Provides context about extraction requirements for conversions

Technical improvements:
- Refactored long lambda expressions into named functions for readability
- Added proper type annotations to avoid type inference conflicts
- Fixed line length violations and whitespace issues
- Added C901 exception for app.py UI functions (complex by nature)
- All quality checks passing (ruff, mypy, tests)

Impact:
- Users can now easily reuse previous work across all major workflow pages
- Reduces manual path entry and potential errors
- Improves workflow integration between different migration stages

* fix: resolve datetime formatting errors and add missing effort estimates

Changes:
1. Fixed TypeError: 'datetime.datetime' object is not subscriptable
   - Changed from created_at[:10] to created_at.strftime('%Y-%m-%d')
   - Applied to all 4 history selection dropdowns (Migration Planning,
     Dependency Mapping, Validation Reports Analysis/Conversion)
   - Added hasattr check for backwards compatibility with string dates
   - Refactored format functions to avoid line length violations

2. Added missing estimated_hours_with_souschef to AI analysis results
   - Updated _build_cookbook_result() to calculate both estimates
   - Converts estimated_effort_days to hours (days * 8)
   - Calculates AI-assisted hours (50% reduction)
   - Ensures parity with rule-based analysis using EffortMetrics

Technical details:
- Split long datetime format expressions across multiple lines
- Used for loops instead of next() for better readability
- Added proper fallback values for edge cases
- All quality checks passing (ruff, mypy, tests)

Impact:
- History selection now works in Migration Planning and Validation Reports
- Manual and AI-assisted estimates now display correctly in analysis results
- Consistent effort metrics across AI and rule-based analysis paths

* feat: add activity breakdown details justifying effort estimates

* feat: add writing vs testing breakdown to activity details

Changes:
1. Enhanced activity breakdown display to show time split:
   - Writing/Conversion time (60% of total)
   - Testing/Validation time (40% of total)
   - Shown for both Manual and AI-assisted scenarios

2. Updated details table with granular columns:
   - Manual Write: Time spent converting without AI
   - Manual Test: Time spent testing/validating without AI
   - AI Write: Time spent converting with AI assistance
   - AI Test: Time spent testing/validating with AI
   - Total Saved: Combined time saved across both activities
   - Efficiency: Overall percentage improvement

Example breakdown for Recipes (4 items):
- Manual Migration:
  * Writing/Conversion: 8.8h (60%)
  * Testing/Validation: 5.9h (40%)
  * Total: 14.7h
- With AI Assistance:
  * Writing/Conversion: 3.1h (60%)
  * Testing/Validation: 2.0h (40%)
  * Total: 5.1h
- Saved: 9.6h (65%)

This provides transparent justification of time estimates by showing:
- What portion is hands-on conversion work
- What portion is testing/validation effort
- How AI reduces both writing AND testing time

Impact:
- Users understand the dual nature of migration effort
- Clear visibility into where time is spent
- Justifies realistic effort estimates with detailed breakdown

* feat: enhance activity breakdown with writing/testing hours and ratios

* feat: update version to 4.1.0 and add activity type constant for custom resources

* feat: improve datetime handling and add validation for cookbook path in dependency analysis

* feat: enhance tar extraction safety with built-in security filtering for path traversal protection

* feat: enhance security in tar extraction and clarify test data with NOSONAR comments

* Potential fix for code scanning alert no. 647: Arbitrary file write during tarfile extraction

Co-authored-by: Copilot Autofix powered by AI <62310815+github-advanced-security[bot]@users.noreply.github.com>

* Potential fix for code scanning alert no. 647: Arbitrary file write during tarfile extraction

Co-authored-by: Copilot Autofix powered by AI <62310815+github-advanced-security[bot]@users.noreply.github.com>

* Fix CLI test failures by reorganizing configure_migration command structure

- Separated Click-decorated command function from pure helper function
- Helper now takes 6 parameters: deployment_target, migration_standard, inventory_source, validation_tools, python_version, ansible_version
- Command function now handles Click parameters (output, interactive) and calls helper
- Fixes 5 failing CLI tests that were receiving unexpected 'output' parameter
- All 1758 tests now pass (up from 1753)

* refactor: remove unused migration_config parameter and update monkeypatching in CLI tests

* security: Fix fork execution risk in release workflow

- Add explicit repository check to prevent execution on forks
- Hardcode repository references instead of using ${{ github.repository }}
- Update all checkout steps to explicitly use 'kpeacocke/souschef'
- Add concurrency group to prevent parallel executions
- Fixes security issue: Checking out code from fork in privileged context

* security: Remove duplicate concurrency block with dynamic reference

- Remove redundant concurrency block that was causing SonarQube flag
- Keep only the hardcoded 'releases' group for security
- Fixes: Checking out code from fork in privileged workflow context

* security: Enhance archive handling with size validation to prevent resource exhaustion attacks

### CODE REVIEW

File: souschef/storage/database.py [hunk 1]
```diff
@@ -1117,6 +1160,56 @@ def get_conversions_by_analysis_id(
...
+
+        """
+        try:
+            with self._connect() as conn:
+                sql = self._prepare_sql("DELETE FROM conversion_results WHERE id = ?")
+                cursor = conn.execute(sql, (conversion_id,))
+                conn.commit()
+                return bool(cursor.rowcount and cursor.rowcount > 0)
+        except Exception:
+            return False
```
[the remark this answers is not in the document]
Comment (kpeacocke) [reply souschef/storage/database.py]: All bare except clauses have been addressed with specific exception handling and explanatory comments, and the code passes all linting checks.

File: souschef/ui/pages/cookbook_analysis.py [hunk 2]
```diff
@@ -1672,6 +1708,81 @@ def _build_cookbook_result(cb_data: dict, assessment: dict, status: str) -> dict
...
+                            else a.get("description")
+                        ),
+                        "complexity_factor": (
+                            a.complexity_factor
+                            if hasattr(a, "complexity_factor")
+                            else a.get("complexity_factor")
+                        ),
+                    }
+                )
+            activity_breakdown = activity_list
```
[the remark this answers is not in the document]
Comment (kpeacocke) [reply souschef/ui/pages/cookbook_analysis.py]: Reused the existing serializer and extended it to include the writing/testing fields so the manual mapping block is gone. The new helper handles objects and dicts in one place to reduce future drift, and _build_cookbook_result now calls the shared serializer.

File: souschef/ui/pages/history.py [hunk 3]
```diff
@@ -137,6 +137,18 @@ def _display_analysis_details(analysis) -> None:
...
+    st.divider()
+
+    # Display activity breakdown if available
+    try:
+        analysis_data = json.loads(analysis.analysis_data)
+        activities = analysis_data.get("activity_breakdown", [])
+        if activities:
+            _display_analysis_activity_breakdown(activities)
+            st.divider()
+    except (json.JSONDecodeError, AttributeError):
```
[the remark this answers is not in the document]
Comment (kpeacocke) [reply souschef/ui/pages/history.py]: Updated the comment on the except block to explain the best‑effort behaviour and why skipping is acceptable in this case.
